build.py

```diff
@@ -5,245 +5,394 @@
     ]
 
-    if ort_home:
-        ort_home = os.path.abspath(ort_home)
-        if not os.path.isdir(ort_home):
-            raise RuntimeError(f"ORT_HOME '{ort_home}' does not exist.")
-        command += [f"-DORT_HOME={ort_home}"]
+    if args.ort_home:
+        command += [f"-DORT_HOME={args.ort_home}"]
 
-    cuda_compiler = None
-    env = {}
-    if cuda_home:
+    if args.use_cuda:
         cuda_arch = 80
-        env["CUDA_HOME"] = cuda_home
-        env["PATH"] = os.path.join(env["CUDA_HOME"], "bin") + os.pathsep + os.environ["PATH"]
-        cuda_compiler = os.path.join(env["CUDA_HOME"], "bin", "nvcc")
-        command += [f"-DCMAKE_CUDA_COMPILER={cuda_compiler}", f"-DCMAKE_CUDA_ARCHITECTURES={cuda_arch}"]
-
-    run_subprocess(command, env=env).check_returncode()
-    make_command = ["cmake", "--build", ".", "--config", config]
-    if parallel:
+        cuda_compiler = str(args.cuda_home / "bin" / "nvcc")
+        command += [f"-DCMAKE_CUDA_COMPILER={cuda_compiler}",
+                    f"-DCMAKE_CUDA_ARCHITECTURES={cuda_arch}"]
+
+    if args.android:
+        command += [
+            "-DCMAKE_TOOLCHAIN_FILE="
+            + str((args.android_ndk_path / "build" / "cmake" / "android.toolchain.cmake").resolve(strict=True)),
+            f"-DANDROID_PLATFORM=android-{args.android_api}",
+            f"-DANDROID_ABI={args.android_abi}",
+            f"-DENABLE_PYTHON=OFF"
+        ]
+
+    util.run(command, env=env)
+
+
+def build(args: argparse.Namespace, env: dict[str, str]):
+    """
+    Build the targets.
+    """
+
+    make_command = [str(args.cmake_path), "--build", str(args.build_dir), "--config", args.config]
+
+    if args.parallel:
         make_command.append("--parallel")
-    run_subprocess(make_command, cwd=build_dir, env=env).check_returncode()
 
-    if not skip_csharp:
-        if not is_windows():
-            warnings.warn('C# API is only supported on Windows.', UserWarning)
-            return
+    util.run(make_command, env=env)
 
-        dotnet = resolve_executable_path("dotnet")
-        configuration = f"/p:Configuration={config}"
-        platorm = "/p:Platform=Any CPU"
-        # Tests folder does not have a sln file. We use the csproj file to build and test.
-        # The csproj file requires the platform to be AnyCPU (not "Any CPU")
-        platform_for_tests_csproj = "/p:Platform=AnyCPU"
-        native_lib_path = f"/p:NativeBuildOutputDir={os.path.join(build_dir, config)}"
+    if util.is_windows() and not args.skip_csharp:
+        dotnet = str(_resolve_executable_path("dotnet"))
 
         # Build the library
-        properties = [configuration, platorm, native_lib_path]
         csharp_build_command = [dotnet, "build", ".",]
-        run_subprocess(csharp_build_command + properties, cwd=os.path.join("src", "csharp")).check_returncode()
+        csharp_build_command += _get_csharp_properties(args)
+        util.run(csharp_build_command, cwd=str(REPO_ROOT / "src" / "csharp"))
```

Comment:
nit: the `util.run` type annotation says `cwd` is `Path | None`

Reply:
Updated type annotation to match what subprocess run supports (it calls os.fsdecode on the input) and removed unnecessary conversion to string.